### PR TITLE
Add a dropdown to enable tag sign

### LINKED ISSUE #3842: [Feature] Tag dialog allow to sign the tag
**Do you want to request a *feature* or report a *bug*?**
It's a feature about tag command.

**What is the current behavior?**
"Create tag" dialog window allows to create an annotated tag but not a signed tag.

**What is the expected behavior?**
Create a signed tag, like already done with commit.

**Which versions of GitExtensions, GIT and Windows are affected by this issue? Did this work in previous versions of our tool?**
- GitExtension 2.50

### PR DESCRIPTION
**Pull request by gpongelli**
Fixes #3842  .

Changes proposed in this pull request:
 - New checkbox into "Create tag" dialog
 - git tag command that depends on this checkbox
 
 
Screenshots before 

image

Screenshot after 
image

How did I test this code:
 - Open "Create tag" dialog
 - Create an annotated tag as usual
 - Verify that the tag is not signed `git tag -v <tagname>`
 - Open "Create tag" dialog again
 - Create a tag with "signed" checkbox enabled
 - Verify that the tag is signed

Has been tested on (remove any that don't apply):
 - GIT 2.11 and above
 - Windows 7 and above
- GitExtension 2.50

### CODE REVIEW

**Comment by RussKie** [comment]
* 'Message' label looks misaligned on the 2nd screenshot
* Personally I'd reword to 'Sign with GPG'
* Allow signing only if there is GPG key 
* Need to provide ability to specify keyid in event there are several keys available (`-u keyid` switch). 
* The checkboxes have to be converted either into radiobuttons or make a dropdown with option since annotation and signing are mutually exclusive (https://git-scm.com/docs/git-tag). 
I prefer the dropdown since it can create a simpler UI. It will have the following options:
  - Annotate
  - Sign with default GPG (visible if keys available)
  - Sign with specific GPG (visible if multiple keys available, makes the scond dropdown visible)
Second dropdown with all available keys

**Comment by gpongelli** [comment]
@RussKie I think a dropdown can be a good choice.
About its content, ok for "Annotate" and "Sign with default GPG" (that will use the config user.signingkey).
But how is it possible to check if multiple keys are available ? 
The config key user.signingkey can be only one.

Could be a solution to enable a textbox on the right of the dropdown to let the user manually enter the specific key id ?
Something like
- Annotate
- Sign with default GPG
- Sign with specific GPG -> enables a textbox to manually enter the key

What do you think?

**Comment by RussKie** [comment]
wrt: enumerating the keys - I don't know 😊 
I had that idea when read through git-tag help, it may not be possible.
A textbox next to the dropdown will work as well.

**Comment by gpongelli** [comment]
Look at the changes, now the tag kind is choosed from a dropdown.

image

image

image

**Comment by RussKie** [comment]
Nice.

* What is "lightweight tag"?
* Should "Message" textbox be only available when "annotated tag" option is selected?

**Comment by gpongelli** [comment]
Lightweight is a not annotated, nor signed tag. The same as actual disabled checkbox.


Quoted from git scm page:

If one of -a, -s, or -u <keyid> is passed, the command creates a tag object, and requires a tag message. 


For this reason I've leaved the message box enabled also when signing the tag.


Ing. Gabriele Pongelli


Il giorno 15 lug 2017, alle ore 06:44, RussKie <notifications@github.com> ha scritto:

**Comment by gpongelli** [comment]
@RussKie also squashed this branch.

File: GitCommands/Git/GitModule.cs [hunk 1]
```diff
@@ -1508,11 +1508,34 @@ public string FormatPatch(string from, string to, string output)
...
+                case 3:
+                    strCommand = $"tag \"{tagName.Trim()}\" -u {keyId} {(force ? " - f" : "")} -F \"{GetGitDirectory()}\\TAGMESSAGE\" -- \"{revision}\"";
+                    break;
+
+                /* Lightweight */
+                default:
+                    strCommand = $"tag {(force ? "-f" : "")} \"{tagName.Trim()}\" \"{revision}\"";
+                    break;
+            }
+
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs]: The basic implementation looks ok, however there few problems with it.

1. All input parameters must be validated. The method is public and you can't control what is being passed in to it.
I don't suppose either `tagName` or `revision` can be null or empty and, if they are, the method must throw `ArgumentNullException`.
`keyId` must be validated for `case 3` and throw if it isn't supplied.

2. There are multiple repeations (e.g. `tagName.Trim()` or `GetGitDirectory()` or `force ? "-f" : ""`) which should be done once.

3. Similar to my suggestion on your other PR, I would extract this functionality in a separate controller (e.g. TagController) and have `IGitModule` passed as a dependency.
Then you can write unit tests which every code path in your method.
For example:

```
        private const string TagName = "bla";
        private const string Revision = "0123456789";
        private const string WorkingDir = "C:\\Working\\Folder";

        [SetUp]
        public void Setup()
        {
            _module = Substitute.For<IGitModule>();
            _module.GetGitDirectory().Returns(x => WorkingDir);

            _controller = new GitTagController(_module);
        }

        [TestCase(true)]
        [TestCase(false)]
        public void Tag_sign_with_default_gpg(bool force)
        {
            var result = _controller.Tag(TagName, Revision, force, 3, "foo");

            _module.Received(1). RunGitCmd($"tag \"{TagName}\" -s {(force ? " - f" : "")} -F \"{WorkingDir}\\TAGMESSAGE\" -- \"{Revision}\""));
        }
```

File: GitCommands/Git/GitModule.cs [hunk 2]
```diff
@@ -1508,11 +1508,34 @@ public string FormatPatch(string from, string to, string output)
...
-        public string Tag(string tagName, string revision, bool annotation, bool force)
-        {
-            if (annotation)
-                return RunGitCmd(string.Format("tag \"{0}\" -a {1} -F \"{2}\\TAGMESSAGE\" -- \"{3}\"", tagName.Trim(), (force ? "-f" : ""), GetGitDirectory(), revision));
-            return RunGitCmd(string.Format("tag {0} \"{1}\" \"{2}\"", (force ? "-f" : ""), tagName.Trim(), revision));
+        public string Tag(string tagName, string revision, bool force, int tagIdx = 0, string keyId = "")
+        {
+            string strCommand = "";
+
+            switch (tagIdx)
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs]: Instead of `tagIdx` I'd call this something like `operation` or `operationType`.
I wonder whether it is worth introducing an enum and use it instead of using magic numbers.

File: GitUI/CommandsDialogs/FormCreateTag.cs [hunk 3]
```diff
@@ -77,7 +77,7 @@ private string CreateTag()
                 File.WriteAllText(Path.Combine(Module.GetGitDirectory(), "TAGMESSAGE"), tagMessage.Text);
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCreateTag.cs]: I would move all tag-creation related functionality into the GitTagController - as the consumer of a method you don't need to know or care how your tag is created, all you have to do is to pass all necessary parameters and get your tag back.
Validation of the tag message and writing of the "TAGMESSAGE" file is the responsibility of the controller, not the caller.
The method should really look like this:

```
       private string CreateTag()
        {
            DialogResult = DialogResult.Cancel;

            string revision = commitPickerSmallControl1.SelectedCommitHash;
            if (revision.IsNullOrEmpty())
            {
                MessageBox.Show(this, _noRevisionSelected.Text, _messageCaption.Text);
                return string.Empty;
            }

            var s = _gitTagController.Tag(textBoxTagName.Text, revision, tagMessage.Text, ForceTag.Checked, annotate.SelectedIndex, textBoxGpgKey.Text);

            if (!string.IsNullOrEmpty(s))
                MessageBox.Show(this, s, _messageCaption.Text);

            if (s.Contains("fatal:"))
                return string.Empty;

            DialogResult = DialogResult.OK;
            return textBoxTagName.Text;
        }

```

I know it is the existing code and you haven't written it but I think it is a worthy refactor which is in line with your overall change.

**Comment by gpongelli** [comment]
@RussKie rebased on top of master and added `GitTagController`

File: GitCommands/Git/GitTagController.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        private IGitModule _module;
+
+
+        public enum TagOperation
+        {
+            Lightweight = 0,
+            Annotate = 1,
+            SignWithDefaultKey = 2,
+            SignWithSpecifiKey = 3
+        };
```
Comment (RussKie) [inline GitCommands/Git/GitTagController.cs]: [Minor] I'd move it in the outer namespace.
Also there is no need to specify values except 0.

File: GitCommands/Git/GitTagController.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            _module = module ?? throw new ArgumentNullException("module");
+        }
+
+        /// <summary>
+        /// Create the Tag depending on input parameter.
+        /// </summary>
+        /// <param name="revision">Commit revision to be tagged</param>
+        /// <param name="tagName">Name of tag</param>
+        /// <param name="force">Force parameter</param>
+        /// <param name="operation">Available tag operationType (Lightweight, Annotate, Sign with defaul key, Sign with specific key)</param>
```
Comment (RussKie) [inline GitCommands/Git/GitTagController.cs]: This should read "_The operation to perform on the tag_" or something similar.

File: GitCommands/Git/GitTagController.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                    if(string.IsNullOrEmpty(keyId))
+                    {
+                        throw new ArgumentNullException("keyId");
+                    }
+                    strCommand = $"tag \"{_tagName}\" -u {keyId} {_forced} -F \"{_directory}\\TAGMESSAGE\" -- \"{revision}\"";
+                    break;
+                    
+                /* Error */
+                default:
+                    throw new ArgumentException("Invalid TagOperation");
```
Comment (RussKie) [inline GitCommands/Git/GitTagController.cs]: I think `NotSupportedException` is more appropriate here

File: GitCommands/Git/GitTagController.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            }
```
Comment (RussKie) [inline GitCommands/Git/GitTagController.cs]: I think you can simplify it further. The only differentiator in the command is the switch: `-a`, `-s` and `-u {keyId}`.
So in the `switch` you can set the `tagSwitch` variable and then build the command:

```
command = $"tag \"{_tagName}\" {tagSwitch} {_forced} -F \"{_directory}\\TAGMESSAGE\" -- \"{revision}\"";
```

File: GitCommands/Git/GitTagController.cs [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+            if (operationType > TagOperation.Lightweight)
+            {
+                File.WriteAllText(Path.Combine(_module.GetGitDirectory(), "TAGMESSAGE"), tagMessage);
+            }
+
+            string strCommand = "";
+            string _tagName = tagName.Trim();
+            string _forced = force ? " - f" : "";
+            string _directory = _module.GetGitDirectory();
```
Comment (RussKie) [inline GitCommands/Git/GitTagController.cs]: please use camelCase notation for local variables and the underscore notation for class instance fields (this is contrary to MS guidelines but made popular by JetBrains Resharper).

File: GitUI/CommandsDialogs/FormCreateTag.cs [hunk 9]
```diff
@@ -66,21 +69,13 @@ private string CreateTag()
...
-                    MessageBox.Show(this, _noTagMessage.Text, _messageCaption.Text);
-                    return string.Empty;
-                }
-
-                File.WriteAllText(Path.Combine(Module.GetGitDirectory(), "TAGMESSAGE"), tagMessage.Text);
-            }
-
-            var s = Module.Tag(textBoxTagName.Text, revision, annotate.Checked, ForceTag.Checked);
 
+            GitTagController.TagOperation _tagOperation = (GitTagController.TagOperation) annotate.SelectedIndex;
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCreateTag.cs]: this doesn't look type-safe.
perhaps a better way is to create a method `private TagOperation GetSelectedOperation()` which extracts the selected operation from the dropdown in a type-safe manner.
This will also make `AnnotateDropDownChanged` more readable and maintainable.

File: GitCommands/Git/GitTagController.cs (hunk 7 above)
Comment (gpongelli) [inline GitCommands/Git/GitTagController.cs]: Following [git tag page](https://git-scm.com/docs/git-tag), tag name should be after the options.
Could `command = $"tag {tagSwitch} {_forced} -F \"{_directory}\\TAGMESSAGE\" \"{_tagName}\" -- \"{revision}\"";` be better ?
I'm in doubt with `--` , I don't know what does it means.

Comment (RussKie) [inline GitCommands/Git/GitTagController.cs]: As the long as the command produces the expected result.

No idea about `--`, it was introduced here fc3d88c961dd5d106f3c48ef3df4da70b535107d
I think it may be a left over from some command...
I suppose you can remove it and test all tag creation scenarios.

File: GitCommands/Git/GitTagController.cs (hunk 4 above)
Comment (gpongelli) [inline GitCommands/Git/GitTagController.cs]: All those value are needed by FormCreateTag to know if "specific key id" shall be enabled or not.
You'll see in my next commit.

Comment (RussKie) [inline GitCommands/Git/GitTagController.cs]: Enums are auto numbered by CLR, but as per design guidelines it is advised to set the default one to 0.
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/enum

**Comment by RussKie** [comment]
Looks good!
Could you please add unit tests and it is good to go

**Comment by gpongelli** [comment]
I'm trying to add some test case, how can I execute them ?

**Comment by vbjay** [comment]
https://msdn.microsoft.com/en-us/library/hh270865.aspx

**Comment by RussKie** [comment]
I use Resharper, but you can also use the VS Test Explorer. 
https://www.google.com.au/search?q=vs+2015+run+unit+test

**Comment by RussKie** [comment]
http://www.ncrunch.net/ is good as well, but it cost money

**Comment by gpongelli** [comment]
@vbjay @RussKie thanks, I'm on VS2017 Community, I'm going to execute tests

**Comment by vbjay** [comment]
The test runner package makes it so the test explorer window will work. No
need to buy anything.

**Comment by RussKie** [comment]
I wasn't suggesting to buy ncrunch, just mentioned it as an FYI. 
The beauty of ncrunch that it can run tests in realtime as you type and show the code coverage. WallabyJs does the same for JS.

**Comment by gpongelli** [comment]
Test class just added.

**Comment by RussKie** [comment]
Good work
Please squash it

**Comment by gpongelli** [comment]
@RussKie squashed

**Comment by RussKie** [comment]
@gpongelli could you please rebase on top of the master so all checks are green and then we can merge it

**Comment by RussKie** [comment]
The unit tests have been restructured, please put `GitTagControllerTests.cs` under UnitTests\GitCommandsTests in the location mirroring the location of `GitTagController.cs`
Thanks

**Comment by gpongelli** [comment]
@RussKie just rebased on top of master and moved GitTagControllerTest.

**Comment by RussKie** [comment]
@gpongelli it looks like you've used C# 7 features, which are VS2017 only. We are currently targeting VS2015+.
image

There are 2 errors.
Please downgrade to C# 6 featureset and fix the invalid xml-doc reference.

Also a number of unit tests failed for me:
image

File: UnitTests/GitCommandsTests/Git/GitTagControllerTest.cs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace GitExtensionsTest.GitCommands.Git
+{
+    class GitTagControllerTest
+    {
+        private const string TagName = "bla";
+        private const string Revision = "0123456789";
+        private const string TagMessage = "foo";
+        private const string KeyId = "A9876F";
+        private const string WorkingDir = "C:\\";
```
Comment (RussKie) [inline UnitTests/GitCommandsTests/Git/GitTagControllerTest.cs]: I think to avoid AAE and account for developer's paths, this should point to NUnit's TestDirectory
http://nunit.org/docs/2.6.2/testContext.html

Comment (gpongelli) [reply UnitTests/GitCommandsTests/Git/GitTagControllerTest.cs]: Changed, I've some problem with Test environment on Visual studio, so don't know if it works.

File: UnitTests/GitCommandsTests/Git/GitTagControllerTest.cs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace GitExtensionsTest.GitCommands.Git
+{
+    class GitTagControllerTest
+    {
+        private const string TagName = "bla";
+        private const string Revision = "0123456789";
+        private const string TagMessage = "foo";
+        private const string KeyId = "A9876F";
+        private string WorkingDir = TestContext.CurrentContext.TestDirectory;
```
Comment (RussKie) [inline UnitTests/GitCommandsTests/Git/GitTagControllerTest.cs]: I would have thought you would need to initialise it in `Setup`
